src/library_fs.js

```diff
@@ -1239,6 +1248,9 @@ FS.staticInit();` +
       return stream.stream_ops.mmap(stream, buffer, offset, length, position, prot, flags);
     },
     msync: function(stream, buffer, offset, length, mmapFlags) {
+#if CAN_ADDRESS_2GB
+      offset >>>= 0;
+#endif
```

Comment:
Hmm, why are the above four casts needed? These functions are not directly called from wasm code?

Reply:
They aren't call from wasm directly, yeah, but a user might call `FS.msync(..)` with a 32-bit pointer. Perhaps the pointer arrived from wasm, or perhaps the user did a `| 0` after an add or such.

Reply:
If the pointer arrives from wasm, I think it should be converted right there and then when it arrives, because that value could be used (incorrectly) somewhere else as well, leaving the missing conversion bug around?

Also, here in these functions this kind of cast would hide real errors that might arise from incorrect calls with negative numbers, since negative numbers would be treated as positive. (e.g. a real FS.msync() call with erroneous negative offset)

Talking about using {{{ }}}  syntax, perhaps here would be a good case to use one? Perhaps we should integrate interop with unsigned, string, int64 conversion, e.g. with
```js
malloc__sig: 'pz', // ('p' for pointer, 'z' for size_t)
foo: function() {
  var nb = 10;
  var ptr = {{{ callWasm('_malloc', 10) }}};
}
```
would turn into
```js
malloc__sig: 'pz', // ('p' for pointer, 'z' for size_t)
foo: function() {
  var nb = 10;
  var ptr = _malloc(10) >>> 0;
}
```
in `CAN_ADDRESS_2GB` mode? Also could expand that with
```js
someFunc__sig: 'sU', // ('s' for UTF-8 string, 'U' for uint64_t)
foo: function() {
  var nb = 10;
  var jsString = {{{ callWasm('_someFunc', 10) }}};
  var jsString = {{{ callWasmSig('_someFunc', 'sU', 10) }}}; // alternative form, specify sig directly?
}
```
would turn into
```js
someFunc__sig: 'sU', // ('s' for UTF-8 string, 'U' for uint64_t)
foo: function() {
  var jsString = UTF8ToString(_someFunc(10, 10/4294967296));
}
```
That way JS callers could automate type conversion for different function calls? (also would obviate needing to declare bindings with webidl_binder.py)

Reply:
> If the pointer arrives from wasm, I think it should be converted right there and then when it arrives, because that value could be used (incorrectly) somewhere else as well, leaving the missing conversion bug around?

Do you mean we should instrument all exports for this? Otherwise I'm not sure how to process all incoming pointers - the code calling wasm can be arbitrary code the user wrote in JS. Fixing up the pointers when they arrive in our FS code seems like a safe thing to do.

> Also, here in these functions this kind of cast would hide real errors that might arise from incorrect calls with negative numbers, since negative numbers would be treated as positive. (e.g. a real FS.msync() call with erroneous negative offset)

Maybe I don't see your point here, but I don't think this would hide a bug? It's a pointer, so if someone passes in -10 that's going to be a huge positive number and we'll fail when we try to load or store to it.

If this were an offset (like an offset in a file) then I agree it would be bad to `>>>=` it, but this is a pointer?

> var jsString = {{{ callWasm('_someFunc', 10) }}};

Interesting idea! I'd like to separate that from this PR, but it definitely is worth considering. I am somewhat concered about the cost of needing to write the signature at each call, which makes me think it doesn't improve readability.

If we do think this is a high-risk area and we want to cover all calls into and out of wasm, then I think instrumenting all wasm exports (as mentioned at the top) might be better - we sacrifice no code readability, and it's guaranteed to apply everywhere. However, it has some downsides, like it adds overhead to each call :(

Reply:
Perhaps autoinstrumenting might be an overkill, but a `{{{ callWasm() }}}` utility might be nice.

> I'd like to separate that from this PR, but it definitely is worth considering.

If we agree that it would be nicer to shift right in the wasm->JS boundary rather than in misc utility functions, we can certainly leave that work for later cleanup as well? (although I must say that I do not particularly care about the code in library_fs.js since that is not in our critical path)

> I am somewhat concered about the cost of needing to write the signature at each call, which makes me think it doesn't improve readability.

Yeah, that is a good point.

Reply:
>  think instrumenting all wasm exports (as mentioned at the top) might be better

The issue with autoinstrumenting that comes to mind is if one routes a wasm function directly to another, e.g.

```js
_callWasmFunction2FromJSThatTakesAPtr(_callWasmFunction1FromJsThatReturnsAPtr());
```
there we would not need to shift, but automatic code transformations would not know that.